### DISCUSSION TITLE
v2.0.0-alpha.4

### DISCUSSION THREAD

**Discussion by markerikson** [reduxjs/redux-toolkit#3321] [Announcements]
This is an **alpha release** for Redux Toolkit 2.0.  This release has many changes to our build setup and published package contents, updates the `redux` and `redux-thunk` deps to the latest alphas, updates the `immer` dep to the latest 10.x beta, and **_may_ have breaking changes**.

### ESM/CJS Package Compatibility

The biggest theme of the Redux v5 and RTK 2.0 releases is trying to get "true" ESM package publishing compatibility in place, while still supporting CJS in the published package.

Earlier alphas made changes to the `package.json` contents and published build artifacts in an attempt to get ESM+CJS compat working correctly, but those alphas had several varying compat issues.

We've set up a battery of example applications in the RTK repo that use a variety of build tools (currently CRA4, CRA5, Next 13, and Vite, Node CJS mode, and Node ESM mode), to verify that Redux and Redux Toolkit compile, import, and run correctly with both TS and various bundlers.  We've also set up a check using a custom CLI wrapper around https://arethetypeswrong.github.io to check for potential packaging incompatibilities.

**This release changes the names and contents of the published build artifacts, and the various `exports/module/main` fields in `package.json` to point to those.**

We already tried to point to ESM build artifacts as the default.  That _should_ hopefully be be more consistent now.

**As of this release, we _think_ we have ESM+CJS compat working correctly**, but we ask that the community try out the alphas in your apps and let us know of any compat problems!

> **Note**: The one known potential issue is that TypeScript's new `moduleResolution: "node16"` mode may see a mismatch between the ESM artifacts and the TS typedefs when imported in a Node CJS environment, and [that _may_ allow hypothetically-incorrect import usage.  (See ongoing discussion in https://github.com/arethetypeswrong/arethetypeswrong.github.io/issues/21 .)  In practice, we _think_ that probably won't be a concern, and we'll do further investigation before a final release.

### Dropping UMD Builds

Redux has always shipped with UMD build artifacts.  These are primarily meant for direct import as script tags, such as in a CodePen or a no-bundler build environment.

For now, we're dropping those build artifacts from the published package, on the grounds that the use cases seem pretty rare today.

We do have browser-ready ESM build artifacts included, which already have `process.env.NODE_ENV` compiled away for either development or production behavior.  These can be loaded via a script tag that points to that file on Unpkg.

If you have strong use cases for us continuing to include UMD build artifacts, please let us know!

### Immer 10 Beta

[Immer 10 is now in beta](https://github.com/immerjs/immer/issues/1015).  It has several major changes, including faster perf, dropping ES5 environment support, and switching from a default export to only named exports.

We've updated RTK to depend on `immer@10.0.0-beta.4`.

Performance testing in an artificial RTKQ stress test project showed [significant perf increases when dealing with many RTKQ-connected components loading](https://github.com/reduxjs/redux-toolkit/pull/3320#issuecomment-1494694657), and that was directly due to Immer-powered reducers executing much faster.

While reducers are not _usually_ the bottleneck in React+Redux apps, this looks like it will be a very nice improvement!

## What's Changed

* Fix lint problems and enable linting on CI by @thorn0 in https://github.com/reduxjs/redux-toolkit/pull/2992
* Allow TS isolatedModules flag to be set for safer transpilation by @matmannion in https://github.com/reduxjs/redux-toolkit/pull/2911
* Test published artifacts in CI by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/3213
* Use Git revision in version and add Node CI examples by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/3258
* Add `arethetypeswrong` automated CLI check by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/3294
* remove "alternative" from descriptions of builder callback by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3296
* update tip regarding overrideExisting to match actual behaviour by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3305
* Add `attw` CLI option to treat problems as non-errors by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/3316
* Rework build setup and hopefully fix ESM compat issues by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/3318
* Bump Immer to 10.0-beta by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/3320

**Full Changelog**: https://github.com/reduxjs/redux-toolkit/compare/v2.0.0-alpha.2...v2.0.0-alpha.4

<hr /><em>This discussion was created from the release <a href='https://github.com/reduxjs/redux-toolkit/releases/tag/v2.0.0-alpha.4'>v2.0.0-alpha.4</a>.</em>

**Comment by einarq** [comment]
https://github.com/immerjs/immer/issues/1015#issuecomment-1497846887

**Comment by einarq** [comment]
Also get this TS issue:

![image](https://user-images.githubusercontent.com/212751/230155561-4eb7c4a4-f4be-4f46-b82a-06a58279e73a.png)

**Reply by einarq** [reply]
Also got this problem when building the code (using Webpack 5):

```
169.7 ERROR in ./node_modules/@reduxjs/toolkit/dist/redux-toolkit.modern.mjs 541:15-22
169.7 Module not found: Error: Can't resolve 'process/browser' in '/usr/src/app/node_modules/@reduxjs/toolkit/dist'
169.7 Did you mean 'browser.js'?
169.7 BREAKING CHANGE: The request 'process/browser' failed to resolve only because it was resolved as fully specified
169.7 (probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
```

Which was resolved by this workaround:
https://github.com/react-dnd/react-dnd/issues/3425#issuecomment-1214554950
`resolve: { fallback: { 'process/browser': require.resolve('process/browser'), } }`

**Comment by einarq** [comment]
@markerikson 
This actually fails inside RTKQ as well, when cache is removed automatically after a little while (for me it seems to happen when the window loses focus).
![image](https://user-images.githubusercontent.com/212751/230289113-fcd95686-12fd-47bf-8d7a-0021865e294d.png)


I created a small repro and commented on the Immer 10 thread:
https://github.com/immerjs/immer/issues/1015#issuecomment-1498535646
